### DISCUSSION TITLE
微信图书 HTML 导出脚本（已失效）

### DISCUSSION THREAD

**Discussion by Sec-ant** [Algebra-FUN/WeReadScan#24] [Ideas]
脚本由于滥用已撤下。

**Comment by Algebra-FUN** [comment]
很强！我看了你的js代码，好奇你是如何发现在页面渲染前存在"#preRenderContent"元素并且其中就有实际的文本内容的？之前我一直被canvas渲染文字所困扰……看了你的代码才意识到还可以在渲染前爬去内容。

**Reply by Sec-ant** [reply]
.

**Reply by Algebra-FUN** [reply]
好奇一个问题，就是"#preRenderContent"元素存在的意义？
当初，当我看到微信读书用canvas渲染正文文字作为反爬手段的时候，就误以为正文文字不会直接暴露到dom中，所以就没多想……
微信读书开发者考虑到反爬虫，为何还要将正文暴露在"#preRenderContent"元素中？因此"#preRenderContent"元素在网页中应该起到一个必不可少的作用：img元素的排版？还是其他用途？
我前端经验不足，好奇这个问题，望可以跟你讨论讨论。

**Reply by Sec-ant** [reply]
.

**Reply by Sec-ant** [reply]
.

**Reply by Algebra-FUN** [reply]
> 看到了[另外一篇博客](https://intoli.com/blog/javascript-injection/#:~:text=This%20is%20the%20ideal%20situation%20because%20it%20means%20that%20we%20can%20be%20certain%20that%20any%20modifications%20to%20the%20JavaScript%20environment%20occur%20before%20any%20page%20code%20is%20executed.)，puppeteer 似乎是可以在页面的脚本执行之前注入自定义脚本的，那么自动化就比较有希望了😄

selenium 也是可以的。不过目前你这个方法严重依赖于#preRenderContent 这个元素的存在。如果微信读书开放团队注意到了这一点并做了修改，这个方法就不适用了:worried:

**Reply by Sec-ant** [reply]
.

**Reply by Algebra-FUN** [reply]
> 而且如果真的彻底改掉了，也可以用最终手段 monkeypatch canvas 来重写它的 fillText 或其他方法进而拿到内容

倒逼微信读书开发团队升级发爬虫技术:smiley:

**Reply by champkeh** [reply]
其实只要弄清楚无限debugger的原理，就好分析代码了，所有的js文件都是同样的处理原理，把对应的无限debugger相关代码删掉，然后用overrides覆盖，就能一步一步调试出请求解密的逻辑了。
我最近花了1周左右空余时间研究了下，最终做了这样一个网站 https://weread.deno.dev/，就是完全采用接口的方式，不需要处理dom

**Comment by Algebra-FUN** [comment]
> 不过欢迎仓库主人参考我的方案

我已经将你的js代码中获取#preRenderContent的部分整合进去了😄，详见 https://github.com/Algebra-FUN/WeReadScan/tree/html-variant

**Reply by Sec-ant** [reply]
赞
